### PR TITLE
Java support for logging metrics with x-coordinates

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?
 
This PR introduces an additional `logMetric()` model to MLflow's Java client APIs allowing users to specify the `step` and `timestamp` at which to record each metric value.
 
## How is this patch tested?
 
The PR expands existing unit tests to verify behavioral correctness of the new `logMetric()` API.
 
## Release Notes

Introduces Java client support for logging metrics against a `step` axis.
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [X] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [X] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by smurching — COMMENTED** [review]
Thanks @dbczumar this looks great, will add the LGTM after we add a test for batched logging per offline discussion

File: mlflow/java/client/src/test/java/org/mlflow/tracking/TestUtils.java [hunk 1]
```diff
@@ -29,11 +29,10 @@ public static void assertMetric(List<Metric> metrics, String key, double value)
     Assert.assertTrue(metrics.stream().filter(e -> e.getKey().equals(key) && equals(e.getValue(), value)).findFirst().isPresent());
   }
 
-  public static void assertMetric(List<Metric> metrics, String key, double value, long runStart, long runEnd) {
```
Comment (dbczumar) [inline mlflow/java/client/src/test/java/org/mlflow/tracking/TestUtils.java]: I could not find a place where this was being used. The signature conflicts with the proposed signature for `assertMetric` with timestamps and steps. Finally, this method seeks to enforce timestamp constraints that do not apply to MLflow metrics.

**Review by smurching — COMMENTED** [review]
LGTM after we change https://github.com/mlflow/mlflow/pull/1202/files/d4caf030f593cb51dc4289022fd03eb613523107..d2ec3dd58d7c8b7eba28bc43072b0df9c2f3227e#diff-a4bd882803aeecc1eaebb72f1668228dR32 to assert on `(key, value, timestamp, step)`, thanks @dbczumar!

**Comment by dbczumar** [comment]
> LGTM after we change https://github.com/mlflow/mlflow/pull/1202/files/d4caf030f593cb51dc4289022fd03eb613523107..d2ec3dd58d7c8b7eba28bc43072b0df9c2f3227e#diff-a4bd882803aeecc1eaebb72f1668228dR32 to assert on `(key, value, timestamp, step)`, thanks @dbczumar!

Done!

**Pull request merged**